src/app/DataModelTypes.h

```diff
@@ -0,0 +1,182 @@
+/*
+ *
+ *    Copyright (c) 2020-2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include "DataModelSerialization.h"
+#include <iterator>
+#include <lib/core/CHIPCore.h>
+#include <lib/core/CHIPSafeCasts.h>
+#include <lib/core/CHIPTLV.h>
+#include <lib/core/Optional.h>
+#include <lib/support/CodeUtils.h>
+
+namespace chip {
+namespace app {
+namespace DataModel {
+
+/*
+ * @brief
+ *
+ * This class provides an iteratable decoder of list items within TLV payloads
+ * such that no memory has to be provided ahead of time to store the entirety of the decoded
+ * list contents.
+ *
+ */
+template <typename T>
+class IteratableList
+{
+public:
+    /*
+     * @brief
+     *
+     * This calls stores a TLV reader positioned on the list this class is to manage.
+     *
+     * Specifically, the passed-in TLV reader should be pointing into the list just after
+     * having called `OpenContainer` on the list element.
+     */
+    void SetReader(const TLV::TLVReader & reader) { mReader = reader; }
+
+    size_t size()
+    {
+        CHIP_ERROR err;
+        TLV::TLVReader reader;
+        reader.Init(mReader);
+        size_t count = 0;
+
+        while ((err = reader.Next()) == CHIP_NO_ERROR)
+        {
+            count++;
+        }
+
+        return count;
+    }
+
+    class Iterator
+    {
+    public:
+        using iterator_category = std::input_iterator_tag;
+
+        /*
+         * Initialize the iterator with a reference to a reader.
+         *
+         * This reader should be pointing into the list just after
+         * having called `OpenContainer` on the list element.
+         */
+        Iterator(const TLV::TLVReader & reader)
+        {
+            mStatus = CHIP_NO_ERROR;
+            mReader.Init(reader);
+        }
+
+        /*
+         * Increments the iterator to point to the next list element
+         * if a valid one exists, and decodes the list element into
+         * the internal value storage.
+         *
+         * If an element does exist and was successfully decoded, this
+         * shall return true.
+         *
+         * Otherwise, if any error was encountered OR the end of the list
+         * was reached, this shall return false.
+         *
+         */
+        bool Next()
```

Comment:
Backstory for how we ended up here from the previous design:

Previously, we had this:

```
for (auto iter = list.begin(); iter != list.end(); iter++)
```

In this scheme, as we advance the iterator, we needed a way to compare the iterator with another iterator that signaled the 'end'.

Since we're traversing TLV, there is no way to just do a pointer-comparison to do that, so I had to do a 'trick' to indicate that was at the end of a TLV list by just marking a flag internally in the iterator as 'the end'. So list.end() is literally just returning an iterator with that flag set to 'true'. Further-more, when I compared two iterators, I actually needed to know they were pointing at the 'same list'. To do that, I stashed away the pointer to the original `IteratableList` to serve as the basis for that comparison.

None of this needed anymore, since in the updated scheme, my iterator model looks like this:
```
            auto iter = t.e.begin();
            while (iter.Next())
            {
                auto & item = iter.GetValue();
            }
```

`iter.Next()` now returns false if it has either hit an error, or end-of-list.

And consequently, the application can call `iter.GetStatus()` to then figure out if something went wrong. Importantly, I've made it such that hitting end-of-list isn't actually construed as an 'error'. In that regard, you can trivially do this to correctly flag issues:

```
auto iter = t.e.begin();
while (iter.Next())
{
      auto & item = iter.GetValue();
}

// Only real errors trigger this. EOF doesn't cause an error, and you carry on merrily.
SuccessOrExit(iter.GetStatus());
```

Hope that helps!